### ISSUE TITLE
Support for Watchman Sonic Advanced/Plus

### ISSUE THREAD

**Issue by garethpotter** [merbanan/rtl_433#2306]
Yet another Watchman oil tank monitor, with an apparently different format.

A few samples are at https://github.com/merbanan/rtl_433_tests/pull/440.

I have managed to decode these using URH following the process outlined in https://github.com/merbanan/rtl_433/issues/1333. The signal is clear enough in URH's demodulated view:

<img width="1093" alt="image" src="https://user-images.githubusercontent.com/108409039/211536243-4e43ef35-f33a-40c3-bc6d-00a541b5f6ed.png">

...but I am not sure how to translate that process into a rtl_433 flex decoder. Running the captures through `rtl_433 -A -a 4` doesn't yield much of use, at least from what I can see.

**Comment by zuckschwerdt** [comment]
That is a pretty neat signal capture. You can even just drop the cu8 onto https://triq.org/pdv/ and see the same curve.
It's GFSK, which rtl_433 isn't optimized for. But the `minmax` demod can pick that up somewhat, e.g.
`rtl_433 -Y minmax -A g100_433.92M_1024k.cu8` -- the timings are off though, actually it's 500 µs PCM like you manually found. E.g. should work, but does not: `rtl_433 -Y minmax -X 'n=name,m=FSK_PCM,s=500,l=500,r=9000'`

We really need someone to build a GFSK-optimized demod someday.

**Comment by garethpotter** [comment]
> We really need someone to build a GFSK-optimized demod someday.

I suspect this is well beyond my capabilities but I will have a dig around.

**Comment by zuckschwerdt** [comment]
The code is here: https://github.com/merbanan/rtl_433/blob/master/src/pulse_detect_fsk.c#L158
I suspect that setting FSK_EST_SLOW to up to 10x the value might help: https://github.com/merbanan/rtl_433/blob/master/src/pulse_detect_fsk.c#L23

**Comment by rolandshiner** [comment]
I just got a Kingspan Watchman Sonic Advanced which I was hoping would work with a rtl-433, but like others here I find sadly it doesn't.

I'm pretty new to rtl-sdr etc but I've been sniffing the airways and have found some useful data. I think I have an ID and both temperature and height but there is other data here I haven't understood yet. The below very long link is my test data on bitbench. *I'm not sure how best to link to that.

The heights shown are always within 1-2 cm of the bottom of the device and the floor. I'm pretty confidant the height data is correct. 

The temperature in the room I'm testing in is approx 18.5C, but it's possible it's reading a little low and the 15.4C shown may be because I put the device on the cold stone floor while adjusting the tripod I am hanging it from. I will try to test the temperature idea further.

https://triq.net/bitbench#c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20c0%2068%20aa%2070%201c%2001%2005%2003%2000%20ad%2083&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20c0%206a%20aa%2070%201b%2001%2005%2003%2000%200c%20e4&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20c0%206a%20a9%2070%2019%2001%2005%2003%2000%208f%2027&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20d8%2068%209a%2070%2014%2001%2005%2003%2000%2048%2040&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20d8%2068%209a%2070%2012%2001%2005%2003%2000%2049%2050&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20d8%2068%20a8%2060%2038%2001%2005%2003%2000%20f4%2052&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20d8%2068%20a9%2060%2038%2001%2005%2003%2000%2075%2041&c=aa%20aa%20aa%20aa%20aa%202d%20d4%200e%2004%2001%2050%204a%2024%20d8%2068%20aa%2070%2055%2001%2005%2003%2000%2065%2038&f=PRE%3Ahhhhhhhhhh%20ID%3Ahhhhhhhhhhhhhhhhhh%20%3F%3F%3Abbbbbbbb%20TEMP-nn.nC%3F%3Ad%20%3F%3F%3Ahh%20HEIGHT-CM%3F%3Ad%20%3F%3F%3Ahhhhhhhh%20%3F%3F%3Add&cw=4

I capture signals using 'rtl_433 -f 433.92M -S unknown'
I use Universal Radio Hacker to analyse them with default settings except for:
Center: -0.52
Samples/Symbols: 125
Signal View: Demodulated
Show Signal as: Hex

I don't know what to do with it from here but maybe others with more experience can take a look.

**Comment by zuckschwerdt** [comment]
Thanks, the `2dd4` is a known sync word, payload starts after that, then the last 16 bits are a CRC-16 poly 0x8005 init 0.

**Comment by garethpotter** [comment]
Right, using my BitBench format string (`40b 8h 8h 8h 8h 8h SERIAL: 24d 8h TEMP: 8b 8h 8b DEPTH: 8d 8h 8h 8h 8h CRC: 24h`) I get a serial number of `05261860` for yours. I note the preamble before the serial number is the same for both devices, though the 8 bits after the serial number differ.

In my case, I used Wireshark to intercept the HTTP POST submissions the USB stick was making (using the firewall to block HTTPS to force it on to HTTP). As I note in the README in my `rtl_433_tests` submission, I am not quite sure how the temperature is encoded, though I know from the HTTP request that it is just a single 8-bit integer (and negative temperatures overflow).

**Comment by rolandshiner** [comment]
Serial number on the device is '5261860'. Clearly they dropped a preceding '0'.

**Comment by zuckschwerdt** [comment]
Combined and clean up the format string is `PRE:40h SYNC:16h MODEL?24h ID:24d ?8h ?8h TEMP:8d ?8h DEPTH-CM:8d 8h 8h 8h 8h CRC:16h`

**Comment by rolandshiner** [comment]
I'm becoming less convinced by my temperature assumption. I'm sometimes seeing it effectively be 026 (2.6C) or 058 (5.8C) while still in the same room at 18.5C.

I'm printing a better tripod mount and then I'll get back to testing

Other than temperature and depth what other information are we expecting to find within the data part of the signal?

**Comment by merbanan** [comment]
Whatever the other watchmen sensors have. But battery status should be somewhere.

**Comment by garethpotter** [comment]
Based on the HTTP submissions, little else. I suspect the temperature is encoded in some way spread across the `8h` _before_ the `TEMP:8d` and potentially afterwards as well, because in my recordings, that `8d` does not change (it is always `170`).

To elaborate, using the following partial format string, `SERIAL: 24d 8h TEMP: 8b 8b 8b DEPTH: 8d`:

`TEMP: 01001001 10101010 01010000` is 0ºC
`TEMP: 01001011 10101010 01100000` is 1ºC
`TEMP: 01001101 10101010 01100000` is 2ºC
`TEMP: 01001111 10101010 01100000` is 3ºC

When I did these recordings it was too cold outside to get higher temperatures! It did actually hit 10º today but I wasn't capturing transmissions.

**Comment by garethpotter** [comment]
There is one other seldom-changing value in the HTTP data. I don't think it is the battery though: it has ranged between 20 and 27 since I started capturing that data at the start of December.

**Comment by zuckschwerdt** [comment]
Other common flags/values include binding, counter, leak/theft alarm
https://github.com/merbanan/rtl_433/blob/master/src/devices/oil_standard.c#L64
https://github.com/merbanan/rtl_433/blob/master/src/devices/oil_watchman.c#L63

**Comment by rolandshiner** [comment]
I can confirm what I had previously thought might be temperature is definitely NOT temperature. I've now got records for 18C, 11C & 8C (all approx)

The unknown 4 bytes between ID and DEPTH in binary are:
11000000  01101000  10101010  01110000 at 18C
11011000  01100001  10101001  01100000 at 11C
11011000  01010100  10101001  01100000 at 8C

@garethpotter Can you see temperature in the app? I don't have the USB receiver.

The basic plug in receiver I have does show low battery so there must be either a voltage level or low battery flag.

I'm pretty sure low level and empty are calculated on the receiver and the device knows nothing about that. It just reports depth in that regard.

Sudden drop is fairly simple to calculate in the receiver but could possibly be checked on the device. I'll try to test for that at some point.

I'll rebind my receiver at some point to look for that data.

I tested for 'NO ECHO CONDITION' but there doesn't appear to be a flag relating to that in the data that I could see but instead I think the warning is just based on chaotic data being received.

**Comment by garethpotter** [comment]
@rolandshiner As far as I know, the temperature is not visible anywhere. I am extracting it as detailed above and dumping it into Home Assistant so I have it there.

I am hoping the temperature will drop below zero over the next 24 hours, so with a bit of luck, I should have some more interesting readings soon.

**Comment by rolandshiner** [comment]
I've been capturing regular, every 30 minute, signals over night to test with rtl_433 using @zuckschwerdt suggested configuration of rtl_433 -Y minmax -X 'n=name,m=FSK_PCM,s=500,l=500,r=9000'. The results are hit and miss. Sometimes it's perfect, sometimes the first part of the data result is correct but presumably an extra detected 1 or 0 then throws off the rest of the data. There are also cases where the data would be correct but it missed the first 1 of the preamble so the data gets distorted accordingly.

I've attached a zip of my test files from overnight. If anyone can use them to further enhance reading the data that would be great, sadly it's well beyond my current understanding.

[watchman advanced test files.zip](https://github.com/merbanan/rtl_433/files/10426746/watchman.advanced.test.files.zip)

**Comment by garethpotter** [comment]
OK, your captures are a bit more useful than mine, presumably because your antenna is closer to the transmitter. I have knocked together a very rudimentary decoder that manages to parse out the serial number and the depth.

https://github.com/garethpotter/rtl_433/commit/0fdd337074d91037fcc3f0140a11529bac1fd03c

It seems to work on most/all of your sample files.

However, I'm still struggling to understand the relationship between the temperature seen in the HTTP POST data and that on the wire. I've had two new values today:

`45 aa 50`, which corresponds to `255`, (i.e. -1º)
`43 a9 50`, `254`, (i.e. -2º)

...but I am still stumped.

**Comment by zuckschwerdt** [comment]
So far we have this list?
```
43a950 [-2 C]
45aa50 [-1 C]
49aa50 [0 C]
4baa60 [1 C]
4daa60 [2 C]
4faa60 [3 C]
54a960 [8 C]
61a960 [11 C]
68aa70 [18 C]
```

A linear regression on the first byte yields `temp C = 0.5205 * Byte - 37.48`
Extreme temperatures would be needed to confirm this and get better coefficients.

**Comment by merbanan** [comment]
Hmm

C=0.55 * F - 32

**Comment by merbanan** [comment]
It could be raw sensor values and then we need a lookup table and/or piece-wise interpolation. I've seen this one time before. We need a lot more readings to figure this out.

**Comment by rolandshiner** [comment]
Allowing for our seperate recorded temperatures to be off by a degree or two in comparison to the temperatures the transmitter recorded:
42 -2 
43 -1.5
44 -1
45 -0.5
46 0
47 0.5
48 1
49 1.5
4a 2
4b 2.5
4c 3
4d 3.5
4e 4
4f 4.5
50 5
51 5.5
52 6
53 6.5
54 7
55 7.5
56 8
57 8.5
58 9
59 9.5
5a 10
5b 10.5
5c 11
5d 11.5
5e 12
6f 12.5
60 13
61 13.5
62 14
63 14.5
64 15
65 15.5
66 16
67 16.5
68 17
69 17.5
6a 18

* It's possible that the usb receiver rounds off the temperature

**Comment by merbanan** [comment]
This is linear with an offset of 0x46.

**Comment by rolandshiner** [comment]
Exactly. Nice and simple.

**Comment by merbanan** [comment]
I think we can let @garethpotter to complete his decoder. It looked fairly ok from a brief look.

**Comment by garethpotter** [comment]
I've tweaked the temperature calculation accordingly. Not had any luck calculating the CRC yet.

A few more temperatures to round out the data:

`3c aa 40` maps to `250`
`3e a9 40` maps to `251`

**Comment by garethpotter** [comment]
And one more overnight:

`3a aa 40` is `249`

**Comment by rolandshiner** [comment]
@garethpotter so using that data I realise 0x45 is 0C and not 0x46.

In summary:
The USB receiver formatting is a signed byte in whole degrees C.
The transmitter formatting is half degrees C with an 0x45 offset.

**Comment by zuckschwerdt** [comment]
> Not had any luck calculating the CRC yet.

The last 16 bits are a CRC-16 poly 0x8005 init 0. The check would be something like
```
        if (crc16(b, 17, 0x8005, 0) != 0)
            return DECODE_FAIL_MIC;
```

**Comment by garethpotter** [comment]
Right, this is done now. I've submitted a PR (https://github.com/merbanan/rtl_433/pull/2323).

**Comment by garethpotter** [comment]
@rolandshiner Should I add some of your samples to the test repository or do you want to do that? They are more useful than mine as they can actually be decoded.

**Comment by garethpotter** [comment]
Right, that's done.

**Comment by Mobbsey** [comment]
Hi,

I've updated rtl_433 to include the changes in #2323 (which btw, thank you for this amazing contribution!), but I don't get any readings coming from my Watchman Sonic Advanced... If I run rtl_433 with -vv, I can see the following screenshot. It seems that it is selecting the correct decoder, but never manages to decode the message. 

The full command I used to get the screenshot below is: `rtl_433 -vv -R 234 -Y minmax`

I'm no expert on rtl_433, but wondered if there was anything obvious I'm missing? Any help would be apprecaited.

![Screenshot 2023-02-09 142437](https://user-images.githubusercontent.com/46693211/217839692-db13994c-f5ca-4ef2-a17d-77116ef6ff5a.jpg)

**Comment by zuckschwerdt** [comment]
Can you grab more of those codes and  copy/paste as text? It looks like bad reception though.

**Comment by zuckschwerdt** [comment]
Can you grab more of those codes and  copy/paste as text? It looks like bad reception though.

**Comment by garethpotter** [comment]
What I’ve found is that if the signal is too strong, it doesn’t work. At the moment, I’m using a gain setting of 36.4 with the antenna inside a shed about 5m from the transmitter. That is decoding reliably every 30 minutes.

If you still don’t have any luck, I suggest trying to capture the signals with `-S unknown` (best to do that with all the default encoders enabled, otherwise you’ll end up with a load of unknown knowns, as Dick Cheney might have it) and then loading them in to URH. That at least would give you a sequence of bits to put in BitBench so we can see if any of the things we thought were constant actually turn out not to be.

**Comment by garethpotter** [comment]
Would also be good to know the serial number of your device so we know when we've found it.

**Comment by Mobbsey** [comment]
Thanks for the quick response both. Let me do some more captures over the next couple of days and I'll come back. Interesting re the signal strength - it's indeed sat on the desk at the moment whilst I tested, so I'll resite it back outside and see if that makes a difference. Thanks so much - I'll feedback what I find FYI.

**Comment by Mobbsey** [comment]
Gents, thanks again for your help. I moved the transmitter back outside, and as per your comment @garethpotter, it's worked reliably overnight. Might be worth a note in the README re the signal strength?

Thanks again, hugely appreciated.

**Comment by garethpotter** [comment]
Hurrah.

I can at least add something to the header of the decoder source file. I might need to make a tweak to the temperature calculation as well - am waiting for the weather to warm up to see how accurate it is - so if so, I can submit a PR for both in due course.

**Comment by Mobbsey** [comment]
If you want a second source of logs to validate temperatures etc once the weather's warmed a little, feel free to reach out. Happy to help.

**Comment by arfond** [comment]
Hi. Would this by any chance work for my plus device? Info on it here:
 https://github.com/merbanan/rtl_433/issues/2133

**Comment by zuckschwerdt** [comment]
> Hi. Would this by any chance work for my plus device? Info on it here: #2133

No, your device codes with variable width symbols and bit stuffing vs. this "CC1101" standard coding.

**Comment by wonkosoft** [comment]
Hi.
Firstly - thanks for the work people have put into this. It was perfect timing for me - almost exactly coinciding with when I decided I was going to try and set up monitoring of my two oil tanks using a couple of these devices. As a software engineer I'd have been happy writing this code, but I'm completely new to decoding radio signals, so having someone else do that hard work was awesome.

Okay. I've got a couple of observations that can hopefully help to further the work, that I'll drop below.

**Comment by wonkosoft** [comment]
First observation - the byte directly preceding the temperature (position 7, using 1-based indexing). In my experiments, that has only ever taken 3 distinct values :
0xC0 - this is during the first 20ish minutes when you've synced the configuration with the receiver and the device is transmitting once per second.
0x80 - this is the first one or two transmissions after the when the device seems to be calibrating itself.
0x98 - this is its normal, live value that you'll see on every transmission when the device is up and running

I'm happy to add that as a flag to the output, but thought it would first be worth seeing if anyone had seen any other values here.

**Comment by wonkosoft** [comment]
Second observation - I had my tanks filled the other day. I only had the one device installed at that point. For a period of a day or two afterwards I was unable to get a reading through RTL_433, as if the device had stopped transmitting. 
Remembering that I'm very new to the radio side of this, I did my best to debug the problem. I found that the device was definitely still transmitting. I thought there might be a problem with noise, but I'm not at all convinced by that because I did see very clear signals and, it seemed, the code was trying to decode that but failed.

I came down to it being one of two possible causes, but couldn't investigate further as by the time I got to this point the problem essentially sorted itself out.
The first possibility was that (because of the fill, which was very high in the tank) the device went into its "No echo" mode and that the transmission in that mode is not of the same length as a regular transmission.
The second possibility, that seems to be more likely, is that there is quite a long part of the transmission which is 'low' (so, maybe a collection of zeros) and the demodulator has incorrectly determined that the transmission has ended, and then a new transmission has started.

I've attached two signal snippets in .cu8 
g002_433.92M_250k.cu8 is able to be decoded.
g004_433.92M_250k.cu8 is not able to be decoded.

Wondered if someone who had a better understanding of this stuff can indicate why I can read one but not the other.

[samples.zip](https://github.com/merbanan/rtl_433/files/10833915/samples.zip)

**Comment by wonkosoft** [comment]
Third observation - there's two bytes between the temperature and depth (so positions 9 and 10 using 1-based indexing). I'm convinced that this is either (1) some representation of time taken for the sonar echo to be received or (2) some representation of strength of sonar echo received, one of which would be used in conjunction with the temperature (as speed of sound changes by temp) for the device to calculate distance. I feel it's more likely (2) as, for a period when the tank was very full I was getting a lot of erroneous readings (all over the place) and this value was always inversely proportional to the reported distance. I think basically the fullness of the tank was causing the device to incorrectly use a secondary echo to measure the distance.

**Comment by wonkosoft** [comment]
Last observation - the constant bytes at positions (1-based) 12,13 and 14 (possibly incorporating 15, which is always zero) look exactly like some kind of version number - either hardware or software. Mine would always read 1.5.3.0.

**Comment by zuckschwerdt** [comment]
Thanks, great infos. For the undecoded signal. Run with `-M protocol` on a good signal. You see the protocol number is `234`, now switch that protocol to verbose with `-R 234:vv` and run also on the bad signal. 
You'll observe the code being chopped off like you guessed. There are 19 sequential 0's, so 9500µs gap, which exceeds the 9000 in the decoder. Change the decoder to `.reset_limit = 12500, // allow 24 sequential 0-bit's` and it will work. I'll update that.

**Comment by zuckschwerdt** [comment]
And there is likely a fourth status byte for low battery. I suspect it's a bit field and battery-good might be the first bit 0x80.

**Comment by wonkosoft** [comment]
Blimey. That was a fast response !!

Thanks. I didn't realise you could do that (-R 234:vv) on the protocol switch - that's very useful. And great news on the solution. I hadn't understood that part of the decoder config, and assumed the problem was earlier in the pipeline, i.e. in the initial demodulation of the FSK signal.

And ... ah. Battery. Yes ... before I started exploring the idea of the values at positions 9 & 10 being related to time/power of the sonar, I did wonder whether the second byte (or, in fact, the 3rd nibble as the 4th is always 0) was related to battery power output. But it was very much a "clutching at straws" theory - it was just because my values for the second byte were always either 0x70 or 0x80, and it seems that it's always the lower value in lower temperatures, which could affect battery output. But, yeah .... probably not. Though I'll watch that value over time (I've added those flags to my local code to see what happens to them).

**Comment by wonkosoft** [comment]
Oh, while I'm here (sorry - I'll stop after this one) - I noted in the manual for the device, on the subject of setting dip switches 1 & 2 (sudden drop & waveguide mode) on the receiver it says "Please be aware that it will take up to 24 hours for [these] settings to be updated to the transmitter", which (more than) suggests that the receiver is also an occasional transmitter. It might be interesting to see what else is on that transmission.

**Comment by zuckschwerdt** [comment]
With the format of the data (preamble, sync, len, payload, crc-16) it is very likely that a transceiver like the CC1101 is used. So technically possible, but I wouldn't expect the sensor to receive commands.

**Comment by garethpotter** [comment]
> First observation - the byte directly preceding the temperature (position 7, using 1-based indexing). In my experiments, that has only ever taken 3 distinct values : 0xC0 - this is during the first 20ish minutes when you've synced the configuration with the receiver and the device is transmitting once per second. 0x80 - this is the first one or two transmissions after the sync period when the device seems to be calibrating itself. 0x98 - this is its normal, live value that you'll see on every transmission when the device is up and running

I've only ever seen `0x80`, but then I've never removed my device from the oil tank since it was installed. @rolandshiner's recordings saw both `0xc0` and `0xd8`, never `0x80` (though there aren't that many).

> Third observation - there's two bytes between the temperature and depth (so positions 9 and 10 using 1-based indexing). I'm convinced that this is either (1) some representation of time taken for the sonar echo to be received or (2) some representation of strength of sonar echo received, one of which would be used in conjunction with the temperature (as speed of sound changes by temp) for the device to calculate distance.

I think the first of these two bytes may actually appear in the data sent to the cloud by the USB receiver. Mine did actually spike briefly today, so that will help confirm if it is indeed the same value. As to what it means, unfortunately, the data format has no schema, so I can't comment there, but it's evidently meaningful enough to be sent to the cloud.

> Last observation - the constant bytes at positions (1-based) 12,13 and 14 (possibly incorporating 15, which is always zero) look exactly like some kind of version number - either hardware or software. Mine would always read 1.5.3.0.

Both mine and @rolandshiner's have also only ever read 0x01050300.

**Comment by wonkosoft** [comment]
> I've only ever seen 0x80

Interestingly (or, possibly not) for the first time since that initial sync, both of my sensors have occasionally transmitted 0x80 today. They seem to do it for a couple of sends, and then go back to 0x98. The only thing that's different today from any other day is that we've had a fairly swift drop in temperature to the lowest we've seen since I got these devices installed.

**Comment by wonkosoft** [comment]
Another observation, similar to @garethpotter 's . I did most of my setup and testing with a receiver a few meters away from the tanks, with only a window between it and the transmitters. I've now got another receiver plugged into my HA box, which is a few more meters away and has a couple of very thick stone walls (it's an old, old house) between receiver and transmitter - and I'm missing far fewer transmissions on that than the original. So it does seem that something in the pipeline - whether it's the receiver, the RTL-SDR drivers or whatever - doesn't like it when the signal is too strong.

**Comment by garethpotter** [comment]
Mine remains steadfast at `0x80`.

I've had a look at the byte that comes just before depth (two bytes after temperature): it definitely correlates with the temperature:

<img width="981" alt="image" src="https://user-images.githubusercontent.com/108409039/223695137-69996018-34ce-40f3-9a3a-b503ff0f653b.png">

Still unsure of the byte in between, which mostly stays at `0xaa` for me, though occasionally I have seen `0xa9`.

**Comment by wonkosoft** [comment]
I've just noticed that the code is truncating the temperature to whole numbers, when it's available from the transmitter in half degrees. Any particular reason to be doing that ? Myself I'd prefer to see the actual reading.

**Comment by garethpotter** [comment]
Yes, I think this was the only reason. In any event, it doesn’t quite match - the offset is not quite linear.

**Comment by klohner** [comment]
In that case, any chance the sensor measures degrees Fahrenheit instead of half-degrees Centigrade, and the receiver was doing the conversion to whole degrees Celsius?

**Comment by gdt** [comment]
@garethpotter What is the status and next steps?

**Comment by rolandshiner** [comment]
It's been a while since I had some time to use this, moving house etc, but I notice that I rarely get a signal from my oil tank sensor unless I use the '-Y minmax' setting. That has the downside of meaning I no longer receive the signals from my various weather sensors. For  now I'm just using another rtl device for general signals as well as a third one for 868Mhz Honeywell devices.

Is there an issue in the rtl_433 code meaning the oil tank sensor is not being correctly recognised and decoded or is this expected behaviour?

**Comment by wonkosoft** [comment]
> It's been a while since I had some time to use this, moving house etc, but I notice that I rarely get a signal from my oil tank sensor unless I use the '-Y minmax' setting. 

For what it's worth, I'm the same. I use minmax and autolevel.

Although ... I get varying results. I've got two tanks, so two sensors. I also have two RTL devices, one for "live" work hooked into my Home Assistant, and one for Dev on my PC. At any given time I may find I'm getting good signals on the live, but missing lots on dev or vice versa, or I'm getting good readings from one sensor and not the other, sometimes the good readings being from one sensor on one RTL device and the other sensor on the other. Weather seems to affect the signals adversely, though not in ways you'd expect - I find that on windy (yet dry) days I miss a lot of readings ... and, no,  I'm not hallucinating ... I think. For quite some time I found that having the RTL antenna laid horizontally was the only way to get the majority of signals ... and then that reversed and now I have to have it vertical.

Funny old world.

Oh, relating to a post further up the chain ... I can definitely confirm that the sensors are receivers as well as transmitters. Changing the dip switch for waveguide mode on the regular receiver definitely changes how the sensor functions, as per the instruction manual.

**Comment by rolandshiner** [comment]
Good to know I'm not the only one. I'm happy running multiple devices I just wanted to be sure I wasn't missing something. Time to buy a fourth receiver so I still have one to play around with!

**Comment by markferry** [comment]
Thanks for your excellent work on this so far @garethpotter.

I have two sensors both purchased from myolive.co.uk around 2017.
They're Tekelek transmitters but use the same encoding as Watchman Sonic Advanced.

Both are model ID 0x0106 so needs at least this patch:

```patch
--- a/src/devices/oil_watchman_advanced.c
+++ b/src/devices/oil_watchman_advanced.c
@@ -47,11 +50,11 @@ Data Layout:
 static int oil_watchman_advanced_decode(r_device *decoder, bitbuffer_t *bitbuffer)
 {
     static uint8_t const PREAMBLE_SYNC_LENGTH_BITS = 40;
-    static uint8_t const HEADER_LENGTH_BITS        = 24;
-    static uint8_t const BODY_LENGTH_BITS          = 128;
+    static uint8_t const HEADER_LENGTH_BITS        = 8;
+    static uint8_t const BODY_LENGTH_BITS          = 144;
     // no need to match all the preamble; 24 bits worth should do
     // include part of preamble, sync-word, length, message identifier
-    uint8_t const preamble_pattern[] = {0xaa, 0xaa, 0xaa, 0x2d, 0xd4, 0x0e, 0x04, 0x01};
+    uint8_t const preamble_pattern[] = {0xaa, 0xaa, 0xaa, 0x2d, 0xd4, 0x0e};
 
     unsigned bitpos = 0;
     int events      = 0;
```

But here's the curious thing:
- sensor 0x011a only decodes when using `-Y minmax`
- sensor 0x0296 only decodes when *not* using it (i.e. just -R 234 is sufficient)

Maybe one sensor is GFSK and the other FSK??


0x011a with -R234:vv -Y minmax:
```
time      : @57.484673s
Model     : Oil-SonicAdv ID        : 00000282
Temperature: 14.000      Depth     : 43            Status    : c0            Integrity : CRC
[pulse_slicer_pcm] Watchman Sonic Advanced / Plus
codes     : {194}aaaaaaaaaa2dd40e010600011ac064a8612b01170000f0770
```

0x0296 -R234:vv:
```
time      : @9.978908s
Model     : Oil-SonicAdv ID        : 00000662
Temperature: 13.000      Depth     : 182           Status    : c0            Integrity : CRC
[pulse_slicer_pcm] Watchman Sonic Advanced / Plus
codes     : {239}ffffffffff8555555555516ea07008300014b6031acb05b008b800005190
```

**Comment by zuckschwerdt** [comment]
@markferry thanks for the detailed notes. I've added your model with 27482e3

**Comment by wonkosoft** [comment]
Whilst irrelevant to the decoding, I've got another issue with these devices that's going to annoy me in a couple of weeks time. 

I've got two devices, and as per the manufacturers instructions, I reset them about 15 minutes apart - as they transmit every 30 minutes and you don't want them transmitting at the same time. Problem is that their clocks aren't exactly accurate, and they're drifting. Currently they're transmitting only about 2mins and 15 seconds apart. About a week ago they were about 3 mins and 10 seconds apart. So in about 2ish weeks they're going to be transmitting at the same time. Hopefully the actual time of clashing will be very short-lived. I discovered this whilst trying to find out why one of the sensors was occasionally telling me the oil level was about a half metre underneath the tank (still don't know why it was doing that). They're a never ending source of joy.

**Comment by markferry** [comment]
> But here's the curious thing:
> 
>     * sensor 0x011a only decodes when using `-Y minmax`
> 
>     * sensor 0x0296 only decodes when _not_ using it (i.e. just -R 234 is sufficient)
> 
> 
> Maybe one sensor is GFSK and the other FSK??
> 

I've tested this again and both sensors are visible using `-Y minmax` so I
think the differences were just a fluke of the GFSK approximation.

That said, during a binding test I'm getting no more than 1 in 20 messages decoded correctly.

**Comment by gdt** [comment]
What's the status?  Can we make progress?

**Comment by markferry** [comment]
Blocked by lack of a good GFSK decoding.
There may be parameters I can tweak to minmax.
Right now 1 in 20 messages is nowhere near good enough.

**Comment by markferry** [comment]
Returned to this and tried tweaking some parameters of protocol 234.

With `short_width = long_width = 580` and `reset_limit = 580 * 24` I'm getting significantly better reception.

Here's the number of messages each setting successfully decoded given the same raw capture:

value | messages decoded
----- | ----
500   | 1
510   | 2
518   | 23
520   | 40
530   | 41
540   | 43
560   | 43
580   | 43
600   | 41
610   | 23
615   | 9
620   | 0

There were 45 complete messages in the sample.

Suffice to say I'm much happier with 95% than the 5% I was getting...

**Comment by markferry** [comment]
I've just tested my second sensor and it decodes best at 620µs.

Raw samples (38MB), first the 580µs sensor ID 282, then the 620µ sensor ID 662:

  http://markferry.net/binding.raw.tbz2

**Comment by zuckschwerdt** [comment]
Thanks. Is that a cu8 file? What freq and sample rate?

**Comment by markferry** [comment]
It's just the raw file that `rtl_433 -w` produces. Tarred and bzipped.

**Comment by zuckschwerdt** [comment]
Without other parameters that's the default 433.92M at 250k then. Rename it to `binding_433.92M_250k.cu8` to e.g. view it on https://triq.org/pdv3/
The timing is 500 µs exacatly. The signal is clipping which degrades the FM demod as our frequency detection needs a sinusoid-like signal.
Try to increase the distance or switch to a manual gain and lower it just until the waveform display in pdv does not show any red or orange.

**Comment by markferry** [comment]
whoa, ok. `-M level -g 15` works beautifully.

Thanks for the hint on the clipping @zuckschwerdt !

**Comment by markferry** [comment]
Follow-up, while attempting to fix one of my sensors I accidentally broke off a resonator. Replaced with a Kingspan Watchman Sonic Advanced ([Amazon £90](https://www.amazon.co.uk/dp/B0032Q8Q64?ref=ppx_yo2ov_dt_b_fed_asin_title)) and setup was *flawless*.

The `tl;dr` is:  if you *manually* set the gain for these GFSK sensors they will work.

**Comment by zuckschwerdt** [comment]
Thanks for the update. I'm working on adding a clipping warning for FSK signals into rtl_433.

**Comment by Harpalyke** [comment]
> Follow-up, while attempting to fix one of my sensors I accidentally broke off a resonator. Replaced today with a Kingspan Watchman Sonic Advanced ([Amazon £90](https://www.amazon.co.uk/dp/B0032Q8Q64?ref=ppx_yo2ov_dt_b_fed_asin_title)) and setup was _flawless_.
> 
> The `tl;dr` is: if you _manually_ set the gain for these GFSK sensors they will work reliably.

@markferry - I'm using the same sensor and had it working very intermittently until recently but I think updating something in homeassistant has caused a problem so Im back fiddling with it again 😄 

Is it something along the lines of this you are running to get best results from the watchman sensor?

`rtl_433 -X 'n=Watchman,m=OOK_PWM,s=500,l=500,r=13920' -R 234 -v -F log -M level -g 15`

**Comment by markferry** [comment]
My current config is:

```
pulse_detect minmax
gain 30
settings digital_agc=1
protocol 234
```

Which is equivalent to:

```
rtl_433 -Y minmax -g 30 -t digital_agc=1 -R 234
```

While I have reliable reception from one sensor I actually have two (at different ranges) and haven't been able to get both of them reliable at the same time.

**Comment by Harpalyke** [comment]
> My current config is:
> 
> ```
> pulse_detect minmax
> gain 30
> settings digital_agc=1
> protocol 234
> ```
> 
> Which is equivalent to:
> 
> ```
> rtl_433 -Y minmax -g 30 -t digital_agc=1 -R 234
> ```
> 
> While I have reliable reception from one sensor I actually have two (at different ranges) and haven't been able to get both of them reliable at the same time.

Thanks @markferry. Seems much more stable now, I'm getting much better reception after altering the gain. Missing the odd update but its almost one every hour now.

**Comment by markferry** [comment]
I took a closer look at my unreliable sensor (model: 0x0106, id: 282) and noticed something interesting.

Readings made sense if they were actually **millimetres, not centimetres**.

Two things:

 - I checked this against a physical reading of the surface distance and it lined up.
 - it explains why readings from that sensor, when I had a batch of them, [were very spiky and showed huge consumption](https://snapshots.raintank.io/dashboard/snapshot/tikYVtaCAe0ciqW7syz56lyfiyuDJlG1) (10 times more, one might say).

What's more in the last couple of days I happened to catch the depth readings *wrap around* (0xff -> 0x00).

```
                  temp_c
               status  |    depth (wtf!)
  model-     _id__  |  |        |
        1 2  3 4 5  6  7  8  9 10
2dd40e 0106 00011a 80 4b a6 30 f4 01170000c9020 
2dd40e 0106 00011a 80 51 a6 41 07 01170000bb240
2dd40e 0106 00011a c0 52 a6 61 07 0117000093090
```

First thing to note is that the least significant nibble of byte 9 must be part of the depth reading (I'll submit a PR for it).

I'm was rather hoping one of the other bytes indicates the reporting precision (mm vs cm) alas I can find no correlation so may just be hardcoded in the firmware for the 0x0106 model.

**Comment by wonkosoft** [comment]
> * it explains why readings from that sensor, when I had a batch of them, [were very spiky and showed huge consumption]

Interesting. Like you I get spikes, and occasional values that make little sense. Over a few years I've been refining the software I wrote to track the data and have assumed the following:
1. My sensors aren't fitted into the tank very well, and at certain times - particularly with very full tanks - I think they (one in particular) gets prone to picking up reflected sonar. The values they return are usually 2x or 3x the the actual value that should be returned. 
2. At certain times of year, due to the tracking of the sun, the tanks can change temperature quite significantly (+/-20C) and that can affect the volume of the oil by a surprising amount. By including temperature in the calculation of volume (based on calculating the volume at a standardised 15C) I get more accurate data.
3. As an aside, I actually don't know the true inner shape and dimensions of the tanks and have made assumptions and estimates that definitely cause inaccuracy in volume calculation at the extremes (very full or almost empty).

But I think I'll do some more raw data capture and see if this idea of the 9th byte contributing to the depth value affects the readings I'm getting.

**Comment by markferry** [comment]
@wonkosoft would be good to know the model ID (0x0106 or 0x0401) of your sensor. I suspect the 9th byte will have no impact on 0x0401 devices which report in cm.

**Comment by wonkosoft** [comment]
Ah, yes. It is a 0x0401.

**Comment by wonkosoft** [comment]
As an aside (through some boredom) I've always thought that the block between temperature and depth encodes some more raw measurement of the sonar function - for example time between sending and receiving the ping. Out of curiosity, I just threw a load of data at an AI and spent a bit of time doing some work to correlate. 
The result seems to be that there is correlation there, as long as you take it on a per-device basis - which suggests there is some other per-device calibration or setting, possibly that is also encoded into that value. The correlation shows that (for one of my devices) the values seem to represent ~ 240 "ticks" per cm of depth, and the other at around ~130 per cm. This is assuming the last half byte of that block is, as suggested, part of the depth reading in mm (it is always zero in my data, suggesting that maybe it was reserved for a future when they started rolling out more accurate devices).
[Edit: And I might currently be using those devices in two different modes - one in the pipe/waveguide mode, one in normal, which may account for the difference between them]
Possibly an AI hallucination, combined with a bias in my mind about what that value is ... but I might get a bigger collection of raw data together and try this exercise again.

**Comment by ProfBoc75** [comment]
I did an update on the decoder because this correction introduce an bug https://github.com/merbanan/rtl_433/issues/2306#issuecomment-1784223757 , the body is still 128 bit and not 144.

I also propose an improvement around the CRC validation, because of GFSK introduce an extra bit and shifting the CRC. Look at PR #3526 and let me know

**Comment by merbanan** [comment]
Implemented, closing issue.
